Index: loader/i386/multiboot.c
===================================================================
--- loader/i386/multiboot_mbi.c
+++ loader/i386/multiboot_mbi.c
@@ -152,6 +152,17 @@
 }
 
 static int
+grub_multiboot_load_elfsym (grub_file_t file, void *buffer, grub_uint32_t *data)
+{
+  if (grub_multiboot_is_elf32 (buffer))
+    return grub_multiboot_load_elfsym32 (file, buffer, data);
+  else if (grub_multiboot_is_elf64 (buffer))
+    return grub_multiboot_load_elfsym64 (file, buffer, data);
+
+  return 0;
+}
+
+static int
 grub_multiboot_get_bootdev (grub_uint32_t *bootdev)
 {
 #ifdef GRUB_MACHINE_PCBIOS
@@ -371,6 +382,9 @@
   if (grub_multiboot_get_bootdev (&mbi->boot_device))
     mbi->flags |= MULTIBOOT_INFO_BOOTDEV;
 
+  if (grub_multiboot_load_elfsym (file, buffer, mbi->syms))
+    mbi->flags |= MULTIBOOT_INFO_ELF_SHDR;
+
   grub_loader_set (grub_multiboot_boot, grub_multiboot_unload, 1);
 
  fail:
